evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,360 @@
+        {
+            if let Ok(call_result) = self.evm.call::<Vec<Address>, _, _>(
+                self.sender,
+                invariant_address,
+                func.clone(),
+                (),
+                U256::zero(),
+                Some(abi),
+            ) {
+                selected = call_result.result;
+            } else {
+                warn!("The function targetContracts was found but there was an error querying addresses.");
+            }
+        };
+
+        self.contracts
+            .clone()
+            .into_iter()
+            .filter(|(addr, _)| {
+                *addr != invariant_address &&
+                    *addr !=
+                        Address::from_slice(
+                            &hex::decode("7109709ECfa91a80626fF3989D68f67F5b1DD12D").unwrap(),
+                        ) &&
+                    *addr !=
+                        Address::from_slice(
+                            &hex::decode("000000000000000000636F6e736F6c652e6c6f67").unwrap(),
+                        ) &&
+                    (selected.is_empty() || selected.contains(addr))
+            })
+            .collect()
+    }
+}
+
+/// The outcome of an invariant fuzz test
+pub struct InvariantFuzzTestResult {
+    pub invariants: BTreeMap<String, Option<InvariantFuzzError>>,
+    /// Every successful fuzz test case
+    pub cases: FuzzedCases,
+}
+
+pub struct InvariantFuzzError {
+    /// The proptest error occurred as a result of a test case
+    pub test_error: TestError<Vec<(Address, (Address, Bytes))>>,
+    /// The return reason of the offending call
+    pub return_reason: Reason,
+    /// The revert string of the offending call
+    pub revert_reason: String,
+    /// Address of the invariant asserter
+    pub addr: Address,
+    /// Function data for invariant check
+    pub func: ethers::prelude::Bytes,
+}
+
+impl InvariantFuzzError {
+    fn new(
+        invariant_address: Address,
+        func: &Function,
+        abi: &Abi,
+        result: &bytes::Bytes,
+        inputs: &[(Address, (Address, Bytes))],
+    ) -> Self {
+        InvariantFuzzError {
+            test_error: proptest::test_runner::TestError::Fail(
+                format!(
+                    "{}, reason: '{}'",
+                    func.name,
+                    match foundry_utils::decode_revert(result.as_ref(), Some(abi)) {
+                        Ok(e) => e,
+                        Err(e) => e.to_string(),
+                    }
+                )
+                .into(),
+                inputs.to_vec(),
+            ),
+            return_reason: "".into(),
+            // return_reason: status,
+            revert_reason: foundry_utils::decode_revert(result.as_ref(), Some(abi))
+                .unwrap_or_default(),
+            addr: invariant_address,
+            func: func.short_signature().into(),
+        }
+    }
+}
+
+pub fn invariant_strat(
```

Comment:
lets move all strategy related things to its own file in foundry/evm/src/fuzz/strategies